keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,572 @@
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	Indexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the pods when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicate that all pods wouldn't be terminated.
+	// - TerminateRunning: indicate that running pods would be terminated.
+	WhenTargetAchieved WhenTargetAchievedSuccessPolicy
+}
+
+// WhenTargetAchievedSuccessPolicy specifies the action to be taken on the pods
+// when the job achieved the requirements.
+// +enum
+type WhenTargetAchievedSuccessPolicy string
+
+const (
+	// All pods wouldn't be terminated when the job reached successPolicy.
+	ContinueWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Continue"
+
+	// Running pods would be terminated when the job reached successPolicy.
+	TerminateRunningWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "TerminateRunning"
+)
+```
+
+Moreover, we validate the following constraints for the `onSucceededPods`:
+- exactly one of the fields `indexes` and `count` is specified
+  for a requirement.
+- whether the specified indexes in the `indexes` and
+  the number of indexes in the `count` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `indexes` is under 64Ki.
+- exactly one of the enum `continue` or `terminateRunning` is specified
+in the `whenTargetAchieved`.
+
+In addition, when the `successPolicy` isn't nil, we validate the `onSucceededPods` isn't nil.
+
+### Evaluation
```

Comment:
I would maybe suggest a section or a comment on what happens in the case of matching `PodFailurePolicy` and `SuccessPolicy`.  

Do you apply `FailurePolicy` first?  And I think FailurePolicy can apply on conditions so I think this case could happen.

Reply:
I think we should respect backOffLimit rather than successPolicy. So the number of failed pods exceeds backkOffLimit and the number of succeeded pods meets successPolicy, the Job declares "Failed".

cc: @alculquicondor @mimowo

Reply:
> @kannon92 I think that `PodFailurePolicy` and `SuccessPolicy` don't match since the `SuccessPolicy` is the feature to declare "Job is succeeded" depends only on whether or not pods succeeded, and `PodFailurePolicy` decides whether or not the failure pods count as failed.
> 
> Does it make sense?

Yea, but I wonder if a single pod matches failure policy, we could declare the job as failed but what if enough indexes are successful?

Reply:
> Yea, but I wonder if a single pod matches failure policy, we could declare the job as failed but what if enough indexes are successful?

Does that mean `.spec.completions=1` and `backoffLimit=1`, right? In that case, as I mentioned above, the job would be declared `Failed` since I think we should respect `backOffLimit` rather than `suceesPolicy`.

Reply:
No, I mean if you have just one pod out of all of them that returns something that matches the `PodFailurePolicy`.  Say a index returns a non-zero exit code or matches a condition

Reply:
And yes, I think you should also consider backoffLimit also.

Reply:
> No, I mean if you have just one pod out of all of them that returns something that matches the PodFailurePolicy. Say a index returns a non-zero exit code or matches a condition

Sorry, maybe I could not understand correctly what you wanted to say. 
IIUC, after the PodFailurePolicy is evaluated, `syncJob` calculates whether or not the number of failed pods exceeds backoffLimit. So, I was thinking that we should consider only collisions of backoffLimit and successPolicy. 

https://github.com/kubernetes/kubernetes/blob/389e60fb2e43e7c13d67e4d9d3a95ed4f088917a/pkg/controller/job/job_controller.go#L958-L969

https://github.com/kubernetes/kubernetes/blob/389e60fb2e43e7c13d67e4d9d3a95ed4f088917a/pkg/controller/job/job_controller.go#L799


@kannon92  If I don't understand correctly, could you clarify?

Reply:
> So `PodFailurePolicy` isn't just about counting the backofflimit. You can say that if a pod has a certain exit code, I want to ignore the backofflimit entirely and fail the job.
> 
> https://kubernetes.io/docs/tasks/job/pod-failure-policy/#using-pod-failure-policy-to-avoid-unnecessary-pod-retries

Ah, right. If the `action` is `FailJob`, the situation will occur as you mentioned above.
So, we should evaluate the `PodFailurePolicy`, first.

Thank you for the clarifications!

Reply:
I would add a note in Evaluation maybe for that.  And it may be worth also mentioning that we will evaluate BackoffLimit first before applying success policy?

Reply:
> I would add a note in Evaluation maybe for that. And it may be worth also mentioning that we will evaluate BackoffLimit first before applying success policy?

I agree with you. Thank you for the help.

Reply:
Done.

Reply:
please resolve old threads. I am happy with the conclusion of this.

Reply:
Sure.